### ISSUE TITLE
[BUG] enforce_index_type doesn't raise an exception for invalid index types.

### ISSUE THREAD

**Issue by ltsaprounis** [sktime/sktime#3641]
**Describe the bug**
The enforce_index_type tag is currently not working as intended.
The issue was discovered in #3605. The FourierFeatures transformer should only allow a pandas PeriodIndex.
However if a DateTimeIndex is passed the transformer will not raise an exception (expected behaviour based on extension contract), nor it will convert the index to period index in before it reaches _fit (expected behaviour based on my understanding of the tag's name).

**To Reproduce**

same code as in #3605
```python
from sktime.datasets import load_airline
from sktime.transformations.series.fourier import FourierFeatures
from sktime.utils.plotting import plot_series

df = load_airline()
df.index = df.index.to_timestamp(freq="M")
from sktime.transformations.series.fourier import FourierFeatures
F_transform = FourierFeatures(sp_list = [12], fourier_terms_list=[2]) # should be yearly
sp1 = F_transform.fit_transform(df)

# Convert seasonality to nanoseconds to get correct result
F_transform = FourierFeatures(sp_list = [12*30*24*3600000000000], fourier_terms_list=[2])
sp2 = F_transform.fit_transform(df)
```

**Expected behavior**

An exception should be raised if the tag is not included in the list in the tag.

I would suggest we also add another tag to convert the index to the specified index, obviously, converting back to the original index type for any public "facing" results. This will **really** simplify development in a few cases e.g. LagTransformer.

**Comment by fkiraly** [comment]
> I would suggest we also add another tag to convert the index to the specified index, obviously, converting back to the original index type for any public "facing" results.

Yes, I think this would be great. I was consiering lifting some of the functionality from the `_ProphetAdapter`, but then again there is a limited number of hours per day.

Maybe open a new issue for this? That's a new (useful) feature, above and beyond the current bug.

**Comment by fkiraly** [comment]
I have discovered the reason: `_check_X_y`, where this should happen, does nothing of the sort.
In fact, nothing at all reads the tag anywhere.

This must be the result of a refactor accident, and lack of testing of said functionality.
(probably my fault)

There should be a few lines in `_check_X_y` that, at the end of conversion:
* check whether the converted-to format is `pandas`
* check whether the index type is in the pass list given by the tag, if not raise an error

**Comment by fkiraly** [comment]
a more sophisticated, principled version adds the index type (or equivalent index type for non-pandas objects) as a metadata field returned by `check_is_mtype`, and checks the metadata field against the tag

**Comment by fkiraly** [comment]
Having thought about this abit more:

I would strongly recommend to think carefully about consequences of starting to raise an exception - this may cause estimators that have previously "over-restricted" the index type to fail, so it may break working user code!

The solution I would go with is some kind of coercion. Let's say:

* if `RangeIndex` or integer based index is enforced, the boilerplate should do `reset_index`, and add back the expected index after `predict`
* if `DatetimeIndex` is enforced, the boilerplate could convert integer-based index to days since Jan 1 1970 (as per Unix time convention), and convert back to integer after `predict`. `PeriodIndex` and `DatetimeIndex` are converted to each other.

I've done something quite similar in `Prophet` (which requires `DatetimeIndex`), in the best case it would only be a matter of lifting this to the `BaseForecaster` boilerplate while ensuring it doesn't eat into runtime by accident.
